packages/esm-form-entry-app/src/app/form-data-source/form-data-source.service.ts

```diff
@@ -114,6 +117,56 @@ export class FormDataSourceService {
     );
   }
 
+  public getMostRecentObsValueBefore(date: string, patientUuid: string) {
+    type ObsResult = FHIRResource['resource'] & {
+      conceptUuid: string;
+      dataType?: string;
+    };
+
+    function isUuid(input: string) {
+      return input.length === 36;
+    }
+
+    const dataSource = {
+      obsValues: [],
+    };
+
+    const concepts = this.conceptResourceService.getAllConcepts();
+    concepts.subscribe((concepts) => {
+      const uuids = concepts.map((c) => c.uuid).join(',');
+
+      return this.observationResourceService.getMostRecentObsValues(date, uuids, patientUuid).subscribe((data) => {
+        const observations =
+          data?.entry?.map((entry) => {
+            const observation: ObsResult = {
+              ...entry.resource,
+              conceptUuid: entry.resource.code.coding.filter((c) => isUuid(c.code))[0]?.code,
+            };
+
+            if (entry.resource.hasOwnProperty('valueString')) {
+              observation.dataType = 'Text';
+            }
+
+            if (entry.resource.hasOwnProperty('valueQuantity')) {
+              observation.dataType = 'Number';
+            }
+
+            if (entry.resource.hasOwnProperty('valueCodeableConcept')) {
+              observation.dataType = 'Coded';
+            }
+
+            return observation;
+          }) ?? [];
+
+        const mostRecentObsBefore = observations.filter(
+          (observation) => new Date(observation.effectiveDateTime).getTime() > new Date(date).getTime(),
+        );
```

Comment:
I think doing this filtering here is a mistake. We need to do it on the backend. Since the backend will only return you a single result per concept and patient (that's what the `max=1` specifies) filtering here just means that including a data will make data unavailable.

Reply:
Also, you could simplify this to `new Date(observation.effectiveDateTime) > new Date(date)`...